pkg/sentry/fsimpl/proc/task_files.go

```diff
@@ -369,6 +370,155 @@ func (d *idMapData) Write(ctx context.Context, src usermem.IOSequence, offset in
+	inode := &memInode{task: task}
+	inode.Init(task.Credentials(), linux.UNNAMED_MAJOR, fs.devMinor, ino, perm)
+
+	taskInode := &taskOwnedInode{Inode: inode, owner: task}
+	d := &kernfs.Dentry{}
+	d.Init(taskInode)
+	return d
+}
+
+func (f *memInode) Init(creds *auth.Credentials, devMajor, devMinor uint32, ino uint64, perm linux.FileMode) {
+	if perm&^linux.PermissionsMask != 0 {
+		panic(fmt.Sprintf("Only permission mask must be set: %x", perm&linux.PermissionsMask))
+	}
+	f.InodeAttrs.Init(creds, devMajor, devMinor, ino, linux.ModeRegular|perm)
+}
+
+func (f *memInode) Open(ctx context.Context, rp *vfs.ResolvingPath, d *kernfs.Dentry, opts vfs.OpenOptions) (*vfs.FileDescription, error) {
+	// TODO(gvisor.dev/issue/260): Add check for PTRACE_MODE_ATTACH_FSCREDS
+	// Permission to read this file is governed by PTRACE_MODE_ATTACH_FSCREDS
+	// Since we dont implement setfsuid/setfsgid we can just use PTRACE_MODE_ATTACH
+	if !kernel.ContextCanTrace(ctx, f.task, true) {
+		return nil, syserror.EPERM
+	}
+	fd := &memFD{}
+	if err := fd.Init(rp.Mount(), d, f.task, f, &f.locks, opts.Flags); err != nil {
+		return nil, err
+	}
+	return &fd.vfsfd, nil
+}
+
+// Dont allow SetStat by default.
+func (*memInode) SetStat(context.Context, *vfs.Filesystem, *auth.Credentials, vfs.SetStatOptions) error {
+	return syserror.EPERM
+}
+
+var _ vfs.FileDescriptionImpl = (*memFD)(nil)
+
+// memFD implements file descriptor for /proc/[pid]/mem.
+type memFD struct {
+	vfsfd vfs.FileDescription
+	vfs.FileDescriptionDefaultImpl
+	vfs.LockFD
+
+	offset int64
+	mu     sync.Mutex `state:"nosave"`
+	task   *kernel.Task
+
+	inode kernfs.Inode
+}
+
+// Init initializes memFD.
+func (fd *memFD) Init(m *vfs.Mount, d *kernfs.Dentry, task *kernel.Task, inode kernfs.Inode, locks *vfs.FileLocks, flags uint32) error {
+	fd.LockFD.Init(locks)
+	if err := fd.vfsfd.Init(fd, flags, m, d.VFSDentry(), &vfs.FileDescriptionOptions{}); err != nil {
+		return err
+	}
+	fd.inode = inode
+	fd.task = task
+	return nil
+}
+
+// Seek implements vfs.FileDescriptionImpl.Seek.
+func (fd *memFD) Seek(ctx context.Context, offset int64, whence int32) (int64, error) {
+	fd.mu.Lock()
+	defer fd.mu.Unlock()
+	switch whence {
+	case linux.SEEK_SET:
+	case linux.SEEK_CUR:
+		offset += fd.offset
+	default:
+		return 0, syserror.EINVAL
+	}
+	if offset < 0 {
+		return 0, syserror.EINVAL
+	}
+	fd.offset = offset
+	return offset, nil
+}
+
+// PRead implementts vfs.FileDescriptionImpl.PRead.
+func (fd *memFD) PRead(ctx context.Context, dst usermem.IOSequence, offset int64, opts vfs.ReadOptions) (int64, error) {
+	if dst.NumBytes() == 0 {
+		return 0, nil
+	}
+	// Buffer the read data because of MM locks
+	buf := make([]byte, dst.NumBytes())
+	n, readErr := getMM(fd.task).CopyIn(ctx, usermem.Addr(offset), buf, usermem.IOOpts{IgnorePermissions: true})
+	if n > 0 {
+		_, err := dst.CopyOut(ctx, buf[:n])
+		if err != nil {
+			return 0, syserror.EFAULT
+		}
+	}
+	if readErr != nil {
+		return int64(n), syserror.EIO
```

Comment:
CopyIn may return a nonnegative n and non-nil error, but returning this from a read system call will cause the sandbox to panic. This is because the Linux system call interface can only indicate a success (positive return value) or an error with 0 bytes read (negative return value), so all partial reads/writes must return without error (you can see how this is handled in pkg/sentry/syscalls/linux/error.go).

It looks like the Linux implementation (see https://github.com/torvalds/linux/blob/master/fs/proc/base.c#L857) will copy from the memory manager in a loop until the the read is completed, or return with error if something goes wrong in the meantime (we'd also return n=0). Perhaps something along the lines of:
```
total := dst.NumBytes()
buf := make([]byte, total)
remaining := total
for remaining > 0 {
  n, err := mm.CopyIn(..., buf[:remaining], ...)
  if n == 0 || err != nil {
    return 0, syserror.EIO
  }
  if _, err := dst.CopyOut(ctx, buf[:n]); err != nil {
    return 0, syserror.EFAULT
  }
  remaining -= n
  offset += n
}
return total, nil
```

Likewise for VFS1. Please double-check my logic though, this is kinda tricky so I might be missing something :)

Reply:
I don't think I understand. As far as I can see does [HandleIOErrorVFS2](https://github.com/google/gvisor/blob/a7b7b7b9804e9968c1fed5f7b3849233f585a88b/pkg/sentry/syscalls/linux/error.go#L39) take care of hiding partial write errors. So the sandbox does not panic.
So why would we need a loop here? I think mm.CopyIn already takes care of copying as much as possible, either in [mm.asCopyIn](https://github.com/google/gvisor/blob/a7b7b7b9804e9968c1fed5f7b3849233f585a88b/pkg/sentry/mm/io.go#L169) or [safemem.CopySeq](https://github.com/google/gvisor/blob/a7b7b7b9804e9968c1fed5f7b3849233f585a88b/pkg/safemem/seq_unsafe.go#L276).

I believe the implementation in Linux does only use the loop because it is limited by the size of the intermediate page used for storing the data during the copy runs.

Reply:
For partial writes, HandleIOErrorVFS2 only takes care of a few errors that we might expect (EIO is not included in that list). 

As far as I can tell, Linux's read implementation will never return a partial result due to the loop--either count reaches 0 or an error is encountered. So even if CopyIn or CopyOut returned with a nil error, we need to return (0, EIO/EFAULT) if we did not copy out all of the bytes, which is the equivalent of returning -EIO/-EFAULT in Linux. I'm fine with avoiding the loop if we do the above, but it might be more natural to follow the logic of the Linux implementation. Please let me know if I'm missing something here, though.

Reply:
I'm still not super comfortable with putting a loop there (especially because I find the kernel implementation not very straight forward to read), but I get the issue now.

I would go with something like this
```go
        buf := make([]byte, dst.NumBytes())
        n, readErr := mm.CopyIn(ctx, usermem.Addr(offset), buf, usermem.IOOpts{IgnorePermissions: true})
        if n > 0 {
                _, err := dst.CopyOut(ctx, buf[:n])
                if err != nil {
                        return 0, syserror.EFAULT
                }
                return int64(n), nil
        }
        if readErr != nil {
                return 0, syserror.EIO
        }
        return 0, nil
```
This way we only return a non-negative number of bytes read if we do not throw an error, handle the 0-bytes-read case as well as partial reads correctly.

Reply:
That looks okay, except I think partial reads (and partial writes) should still return 0 with error. If anyone assumes that success implies the read was fully completed (which Linux seems to do), they could end up accessing an illegal memory location past the end of a partial read. 

You need to compare the return values of both CopyIn and CopyOut with len(buf), and if they are not equal, return EIO/EFAULT. This is kind of unnatural, which is why I slightly prefer the loop. If you're inclined to omit the loop, it would be good to add a comment explaining that avoiding a partial result matches the behavior of the Linux implementation.

Reply:
Ok, so I looked up the contract of the IO interface (https://godoc.org/gvisor.dev/gvisor/pkg/usermem#IO).

Relevant part:
```
    // CopyOut copies len(src) bytes from src to the memory mapped at addr. It
    // returns the number of bytes copied. If the number of bytes copied is <
    // len(src), it returns a non-nil error explaining why.
```
So if dst.CopyOut goes wrong and copies some amount of bytes smaller than N, fine we get an error meaning syserror.EFAULT in the end. So why should we compare if N and the amount of bytes "CopyOut"ed are equal?

Regarding partial results: I did not find any part in the original base.c loop where it returns an error (from doing a partial read) if there have been any bytes read by `access_memory_vm` (since `this_len` is always bigger than 0 in this case).

I have a test for it locally, if you do a read that overlaps with an unmapped section but starts in a mapped one, you will get partial results without any error.

Reply:
Ah okay, thanks for digging into this. That's not the behavior I would have expected from looking at the source code, but I guess that's why we have tests :)